RecoHI/HiCentralityAlgos/plugins/HiHFFilterProducer.cc

```diff
@@ -0,0 +1,100 @@
+
+#include <memory>
+
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/StreamID.h"
+
+#include "DataFormats/CaloTowers/interface/CaloTower.h"
+#include "DataFormats/CaloTowers/interface/CaloTowerCollection.h"
+#include "DataFormats/HeavyIonEvent/interface/HFFilterStruct.h"
+#include "Geometry/Records/interface/CaloGeometryRecord.h"
+
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescriptionFiller.h"
+
+#include "DataFormats/Common/interface/Wrapper.h"
+
+#include <cmath>
+
+class HiHFFilterProducer : public edm::stream::EDProducer<> {
+public:
+  explicit HiHFFilterProducer(const edm::ParameterSet&);
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+private:
+  void produce(edm::Event&, const edm::EventSetup&) override;
+  edm::EDGetTokenT<CaloTowerCollection> srcTowers_;
+};
+
+HiHFFilterProducer::HiHFFilterProducer(const edm::ParameterSet& iConfig)
+    : srcTowers_(consumes<CaloTowerCollection>(iConfig.getParameter<edm::InputTag>("srcTowers"))) {
+  produces<pat::HFFilterStruct>("hiHFfilters");
+}
+
+void HiHFFilterProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+  using namespace std;
+  using namespace edm;
+
+  auto const& towers = iEvent.get(srcTowers_);
+
+  int nTowersTh2HFplus = 0;
+  int nTowersTh2HFminus = 0;
+  int nTowersTh3HFplus = 0;
+  int nTowersTh3HFminus = 0;
+  int nTowersTh4HFplus = 0;
+  int nTowersTh4HFminus = 0;
+  int nTowersTh5HFplus = 0;
+  int nTowersTh5HFminus = 0;
+
+  auto HFFilterResults = std::make_unique<pat::HFFilterStruct>();
+  for (const auto& tower : towers) {
+    const auto et = tower.et();
+    const auto energy = tower.energy();
+    const auto eta = tower.eta();
+    const bool eta_plus = (eta > 3.0) && (eta < 6.0);
+    const bool eta_minus = (eta < -3.0) && (eta > -6.0);
+    if (et < 0.0)
+      continue;
+    if (eta_plus) {
+      if (energy >= 2.0)
+        nTowersTh2HFplus += 1;
+      if (energy >= 3.0)
```

Comment:
The cascade of ifs can be moved inside the previous if. (or alternatively one could rewrite without the need of conditionals..)

Reply:
sorry, I did not get your point here. The if on et < 0 is needed, and eta_plus and eta_minus are important for HF+ and HF- towers. Maybe what I can try to do something using switch, or do you have better idea?

Reply:
Maybe David can clarify what he had in mind, but one way to shorten the code is:
```C++
if (eta_plus) {
  nTowersTh2HFplus += energy>2.0 ? 1 : 0;
  nTowersTh3HFplus += energy>3.0 ? 1 : 0;
  nTowersTh4HFplus += energy>4.0 ? 1 : 0;
  nTowersTh5HFplus += energy>5.0 ? 1 : 0;
} else if (eta_minus) {
  nTowersTh2HFminus += energy>2.0 ? 1 : 0;
  nTowersTh3HFminus += energy>3.0 ? 1 : 0;
  nTowersTh4HFminus += energy>4.0 ? 1 : 0;
  nTowersTh5HFminus += energy>5.0 ? 1 : 0;
}
```

Reply:
one can do
if energy > 2.0  {
   if energy > 3.0 {
      if energy > 4.0 {

as energy can't be > 3 unless its > 2.

Alternatively, less calculations to fill an array
loop over stuff
if eta_plus
  eta_plus_e_hist[min(4,int(energy)]++;

then make a cumsum at the end to get the sum for e>2,3,4,5

Reply:
I would prefer not to reimplement a histogram from scratch here, seems like the potential for hard-to-debug off-by-one errors is large. Perhaps just stick with the nested ifs then, rather than the ternary case, to potentially save a few ops?

Reply:
@davidlange6 is that ok to implement the ifs as in @jpata comment?

Reply:
implemented in commit e67d14cd06504aacc3f89f5e6a3f3ea50235d9fa as in @jpata comment